packages/optimizer/lib/transformers/AutoExtensionImporter.js

```diff
@@ -0,0 +1,269 @@
+    this.enabled = config.autoExtensionImport !== 'false' && config.autoExtensionImport !== false;
+    this.log_ = config.log.tag('AutoExtensionImporter');
+
+    // We use the validation rules to infer extension imports. The rules are downloaded once and for
+    // efficency, we initially extract all needed rules
+    this.extensionSpec = validatorRules.fetch().then((rules) => {
+      // Map extension names to more info
+      const extensionsMap = new Map(rules.extensions.map((ext) => [ext.name, {
+        name: ext.name,
+        type: ext.extensionType === 'CUSTOM_TEMPLATE' ? 'custom-template' : 'custom-element',
+        version: ext.version.filter((v) => v !== 'latest'),
+      }]));
+      // Maps tags (e.g. amp-state) to their extension (e.g. amp-bind)
+      const tagToExtensionsMapping = new Map();
+      // Maps tags to their extension specific allowed attributes
+      // (e.g. amp-img => amp-fx => amp-fx-collection)
+      const tagToAttributeMapping = new Map();
+      // Maps tags to their bindable attributes (e.g. div => text)
+      const tagToBindAttributeMapping = new Map();
+
+      // Iterate over all available tags
+      for (const tag of rules.tags) {
+        const tagName = tag.tagName.toLowerCase();
+
+        // Map amp tags to their required extension(s)
+        if (tagName.startsWith('amp-')) {
+          tagToExtensionsMapping.set(
+              tagName,
+              tag.requiresExtension,
+          );
+        }
+        // Hack: fix missing attribute dependencies (e.g. amp-img => lightbox => amp-lightbox-gallery)
+        tag.attrs.filter((a) => manualAttributeToExtensionMapping.has(a.name))
+            .forEach((a) => {
+              a.requiresExtension = [manualAttributeToExtensionMapping.get(a.name)];
+            });
+
+        // Map attributes to tags and extensions (e.g. amp-img => amp-fx => amp-fx-collection)
+        tag.attrs.filter((a) => a.requiresExtension && a.requiresExtension.length > 0)
+            .forEach((a) => {
+              let attributeMapping = tagToAttributeMapping.get(tagName);
+              if (!attributeMapping) {
+                attributeMapping = [];
+                tagToAttributeMapping.set(tagName, attributeMapping);
+              }
+              attributeMapping.push(a);
+            });
+
+        // Map tags to bindable attributes which are named link `[text]`
+        const bindableAttributes = tag.attrs.filter((a) => a.name.startsWith('['))
+            .map((a) => a.name.substring(1, a.name.length - 1));
+        tagToBindAttributeMapping.set(tagName, new Set(bindableAttributes));
+      }
+      return {
+        extensionsMap,
+        tagToExtensionsMapping,
+        tagToAttributeMapping,
+        tagToBindAttributeMapping,
+      };
+    });
+  }
+
+  async transform(tree, params) {
+    if (!this.enabled) {
+      return;
+    }
+    const html = tree.root.firstChildByTag('html');
+    const head = html.firstChildByTag('head');
+    if (!head) return;
```

Comment:
Can there really be no head? Does this also apply to body? Also, why not define body here too?